Resolved the UnsupportedOperationException in TlsConfigUtils when setting ALPN false

amusarra opened the pull request:
Applying the fix to avoid the UnsupportedOperationException in Tls Config Utils when setting ALPN false. For more detail about this, you can see #46652

cescoffier commented on extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/runtime/config/TlsConfigUtils.java:
```diff
@@ -78,7 +78,9 @@ public static void configure(TCPSSLOptions options, TlsConfiguration configurati
             }
             options.setEnabledSecureTransportProtocols(sslOptions.getEnabledSecureTransportProtocols());
-            options.setUseAlpn(sslOptions.isUseAlpn());
+            if (sslOptions.isUseAlpn()) {
```
I'm not sure. The difference with the previous code is that it won't configure ALPN if sslOption.isUseAlpn is set to `false`, meaning that the options will use its default value.
It should be `false` too, so, that change does not bring much. What am I missing?

amusarra replied:
Hi @cescoffier If you read the issue linked to this PR, you will see that with the current code, even setting the ALPN property to false, you get the blocking exception object of the PR, exception thrown by the AMQP client. In this case, how would you solve the problem that is blocking? Do you have any better ideas? 
Thanks.

cescoffier replied:
ProtonOptions need to be fixed.

amusarra replied:
I'll try to check.

amusarra replied:
> ProtonOptions need to be fixed.

But if the Proton client does not actually support setting ALPN it is correct that it throws the exception that the operation is not supported. In my opinion it should be the one who uses the property setting method to handle the exception correctly if agreed?

cescoffier replied:
No, because the day we have an option implementation using "true" as default value (which may already exist), the new logic is not correct.

We could wrap the current code with the try/catch explaining the issue - but it's ugly.

amusarra replied:
so, in your opinion, what would be the solution to modify the Proton client's set method so that it doesn't throw any exceptions?

amusarra replied:
@cescoffier How would you make the change?

<img width="616" alt="image" src="https://github.com/user-attachments/assets/63a1cbed-3eaa-4447-8904-e3ceb4b17cb2" />

amusarra replied:
ALPN is an extension to TLS that allows you to negotiate the application protocol to use during the TLS handshake. This is particularly useful when a server supports multiple protocols on the same port, such as HTTP/2, MQTT, or AMQP. In the context of AMQP over TLS (AMQPS), the use of ALPN is rare but possible,

Yes, it is correct that the `setUseAlpn(boolean useAlpn)` method in `ProtonClientOptions` throws an `UnsupportedOperationException()`, as this clearly indicates that ALPN is not supported in this implementation.

Proton (AMQP Library), does not have native integration with ALPN.

Given these assumptions I think it is correct to enclose the code in a try/catch block. Another possibility, if you do not like this solution, give the possibility to override `TlsConfigUtils` so that then anyone, in situations like this, can override the default behavior.

By opening the issue to ProtonClient, they might reply that the exception returned is correct because ALPN is not supported. So, what strategy do you propose? Consider that this situation is absolutely blocking for those who want to use AMQP over TLS.

Thanks.

cescoffier replied:
The proton options should not throw an exception but log a message. Otherwise, any implementation could decide what they like or not, making interfaces totally useless. I'm actually surprised they do not support ALPN. 

The TlsConfigUtils cannot be overridden, it's mostly static and it's used in the TLS registry - it won't distinguish one case or another.

The try/catch solution is ugly, but I think it's the best we can have (but please raise an issue in Vertx-proton).

amusarra replied:
Hi @amusarra 
Thanks for the feedback. I'll change the PR to include a try/catch block and then open an issue to Vert.x Proton hoping they'll accept.

amusarra committed:
fix: avoid UnsupportedOperationException in TlsConfigUtils when setting ALPN false

cescoffier commented on extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/runtime/config/TlsConfigUtils.java:
```diff
@@ -78,7 +82,15 @@ public static void configure(TCPSSLOptions options, TlsConfiguration configurati
+            try {
+                options.setUseAlpn(sslOptions.isUseAlpn());
+            } catch (UnsupportedOperationException e) {
+                log.warn(
```
Can you use warnf instead?

amusarra replied:
Yes, sorry, you're right. I made the change.

cescoffier reviewed (approved):
Thanks, looks good.

I add a small comment about using warnf instead of ".formatted"

amusarra commented:
Hi @cescoffier
I have open a new feature request to Vert.x vert-x3/vertx-proton/issues/139

cescoffier reviewed (approved):
Thanks!

cescoffier commented:
Can you squash your commits?

amusarra commented:
> Can you squash your commits? 

I can't now. I'll do it later.

amusarra commented:
> Can you squash your commits?

Executed rebase with the upstream and squash commit.

amusarra commented:
Hi @cescoffier should I do the rebase or do you do it?
Given the high frequency of commits, it is possible that after I do the umpteenth rebase and push after a few minutes, it is possible that I will have to do it again.

cescoffier commented:
Squash and rebase - I don't expect conflicts.

amusarra commented:
> Squash and rebase - I don't expect conflicts.

I already squashed it as you can see and rebased it first. But then some other commits came upstream, no conflicts.

amusarra commented:
@cescoffier I just did the rebase from upstream and push.

gsmet commented:
Thanks for the pull request and the discussions!
